### PR TITLE
Baklava: Full coverage of Discord Rich Presence

### PR DESCRIPTION
**Pull request by amitojsingh366**
- Adds full discord rich presence into baklava *(supersedes #1887)*

###### I already asked ben about this and he's given me the green flag.

### CODE REVIEW

File: baklava/src/utils/rpc/ipc.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+                    setPresence({ state: 'Customising voice settings' });
+                    break;
+                case "overlay-settings":
+                    setPresence({ state: 'Customising overlay settings' });
+                    break;
+                case "sound-effect-settings":
+                    setPresence({ state: 'Customising sound effect settings' });
+                    break;
+                default:
+                    setPresence({ state: 'Exploring the home page' });
```
Comment (theotterlord) [inline baklava/src/utils/rpc/ipc.ts]: so if default is for undefined cases, I'd probably just set the state to something generic
```suggestion
                    setPresence({ state: 'Taking DogeHouse to the moon' });
```
or something like that (just a thought)

**Review by TheOtterlord — APPROVED** [review]
:+1:

File: baklava/src/utils/rpc/ipc.ts (hunk 1 above)
Comment (amitojsingh366) [reply baklava/src/utils/rpc/ipc.ts]: 👍

File: baklava/src/utils/rpc/ipc.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+                    setPresence({ state: 'Customising overlay settings' });
+                    break;
+                case "sound-effect-settings":
+                    setPresence({ state: 'Customising sound effect settings' });
+                    break;
+                case "profile":
+                    setPresence({ state: `Viewing ${pageData.data}'s profile` });
+                    break;
+                default:
+                    setPresence({ state: 'Exploring the home page' });
```
Comment (BronzW) [inline baklava/src/utils/rpc/ipc.ts]: Should be `details` not `state`

File: baklava/src/utils/rpc/ipc.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+            }
+            let pdata = {
+                state: `In ${data.currentRoom.room.name}`,
+                partyId: data.currentRoom.room.id,
+                partySize: data.currentRoom.users.length,
+                partyMax: data.currentRoom.users.length,
+                startTimestamp: Date.parse(data.currentRoom.room.inserted_at),
+                smallImageKey: isSpeaker && !isMuted ? 'mic_on' : 'mic_off',
+                smallImageText: isSpeaker ? `Speaker - ${muted}` : `Listener`,
+            }
```
Comment (BronzW) [inline baklava/src/utils/rpc/ipc.ts]: Can you add a join link button?

Comment (amitojsingh366) [reply baklava/src/utils/rpc/ipc.ts]: sure

File: baklava/src/utils/rpc/ipc.ts (hunk 2 above)
Comment (amitojsingh366) [reply baklava/src/utils/rpc/ipc.ts]: it dosen't really matter but ok

**Review by BronzW — APPROVED** [review]
Looks good

**Comment by BronzW** [comment]
If people agree, a private room should not display any information other than saying that they are in a private room right? Gives private rooms a bit more privacy.

**Comment by BronzW** [comment]
And this does not cause any errors when people who don't have installed discord try to run the electron application?

**Comment by TheOtterlord** [comment]
As far as I know, it is handled by the library. I have the library on a side project ([DeckMaster](https://github.com/TheOtterlord/deckmaster)) and it works fine without discord

**Comment by amitojsingh366** [comment]
> 
> 
> And this does not cause any errors when people who don't have installed discord try to run the electron application?

@BronzW iv'e added error handling for that

**Comment by amitojsingh366** [comment]
> 
> 
> If people agree, a private room should not display any information other than saying that they are in a private room right? Gives private rooms a bit more privacy.

What all do you think should be hidden?

**Comment by TheOtterlord** [comment]
> 
> 
> > If people agree, a private room should not display any information other than saying that they are in a private room right? Gives private rooms a bit more privacy.
> 
> What all do you think should be hidden?

Room name, desc, users, everything.
It should just show that your in a private room, or nothing

**Comment by amitojsingh366** [comment]
> 
> 
> > > If people agree, a private room should not display any information other than saying that they are in a private room right? Gives private rooms a bit more privacy.
> > 
> > 
> > What all do you think should be hidden?
> 
> Room name, desc, users, everything.
> It should just show that your in a private room, or nothing

ok done!

**Comment by BronzW** [comment]
I think this is ready now

**Comment by BronzW** [comment]
Does it refetch the new data every 15 seconds?

**Comment by BronzW** [comment]
Does this just get the data once or does it update every few seconds?

**Comment by amitojsingh366** [comment]
> Does this just get the data once or does it update every few seconds?

whenever data updates rpc updates

**Comment by amitojsingh366** [comment]
> I think this is ready now

yea but @benawad said to wait for him to review new features.

**Comment by BronzW** [comment]
@amitojsingh366 Are you aware of the ratelimit?

**Comment by amitojsingh366** [comment]
@BronzW afaik ratelimit is only on the api not the rpc. so even if we get rate limited it wont cause an error for us. it will cause an error from discord's app

**Pull request merged**